Initial Commit

IIExpanse opened the pull request:
Андрей, добрый день!
Отправляю ТЗ-10 на ревью.
Помимо требований ТЗ, в класс Film добавлено поле rate - оно отражает количество лайков и обновляется при добавлении/удалении лайков. Также, если при возврате популярных фильмов нет фильмов с лайками, возвращается весь список целиком.

Всего этого нет в ТЗ, но тесты постмана ждут именно такой логики (консультировались с наставником из-за проблем с тестами). Например, там есть тест, который после удаления единственного лайка ожидает возврат списка популярных фильмов с 2 объектами. Поле rate тесты пока не требуют, но оно присутствует в передаваемых объектах.

IIExpanse committed:
Initial Commit

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,95 @@
+    private final FilmService service;
+    private final UserStorage userStorage;
+
+    @Autowired
```
Тут и далее над конструктором нет необходимости ставить аннотацию, спринг поймет, что для создания контроллера требуются зависимости и внедрит их, а т.к. в зависимостях есть ломбок, то можно код еще сократить и не объявлять явно конструктор

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Autowired
+    public FilmController(FilmStorage storage, FilmService service, UserStorage userStorage) {
+        this.storage = storage;
```
Тут и далее в текущей архитектуре используются слои  сервис, контроллер и слой, отвечающий за хранение - storage. Лучше  их не смешивать, т.е. контроллер знает только о сервисе, а сервис знает о слое хранения. Т.е. контроллеру лучше не знать про storage, а все манипуляции производить с сервисом

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,95 @@
+package ru.yandex.practicum.filmorate.controllers;
```
Лучше привести пакеты к одному стилю, т.е. сделать в единственном числе

IIExpanse committed:
fix: 1. Убраны излишние аннотации @Autowired, конструкторы заменены аннотациями @AllArgsConstructor;
2. Контроллеры вызывают только сервисы. Методы хранилищ продублированы в сервисы (интерфейсы и реализации), откуда происходит вызов аналогичных методов хранилищ;
3. Для избежания проблем с десериализацией к классам Film и User добавлены аннотации @NoArgsConstructor и @AllArgsConstructor.

style: Стиль названий пакетов приведен к наименованиям в единственном числе.

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,64 +1,99 @@
-    private final Map<Integer, Film> filmsMap = new HashMap<>();
-    private int idCounter = 1;
+    private final FilmService service;
+    private ServletWebServerApplicationContext context;
```
Решение в внедрением контекста (т.е. инфраструктуры) и обращение к себе по http, не самое удачное решение. Лучше проверку спрятать в сервис, а сервис может при необходимости обратиться к стораджу с пользоватеями

IIExpanse committed:
fix: Обращение к UserController от FilmController через HTTP для проверки существования пользователя заменено на обращение к FilmService. В InMemoryFilmService для реализации логики проверки добавлено поле с хранилищем пользователей.
